File: libs/core/src/broadcast/core/crypto.py
"""Symmetric encryption via ``gpg``.

The passphrase is handed to gpg over a dedicated pipe fd — never on argv
(where ``ps`` would show it) and never written to disk.
"""

from __future__ import annotations

import os
import subprocess

from broadcast.core.errors import BroadcastError


def _gpg(args: list[str], data: bytes, passphrase: str) -> bytes:
    read_fd, write_fd = os.pipe()
    os.write(write_fd, passphrase.encode() + b"\n")
    os.close(write_fd)
    try:
        proc = subprocess.run(
            [
                "gpg",
                "--batch",
                "--yes",
                "--quiet",
                "--pinentry-mode",
                "loopback",
                "--passphrase-fd",
                str(read_fd),
                *args,
            ],
            input=data,
            capture_output=True,
            pass_fds=(read_fd,),
        )
    except FileNotFoundError as exc:  # pragma: no cover - env-specific
        raise BroadcastError("gpg not found on PATH") from exc
    finally:
        os.close(read_fd)
    if proc.returncode != 0:
        raise BroadcastError("gpg failed: " + proc.stderr.decode(errors="replace").strip())
    return proc.stdout


def encrypt(plain: bytes, passphrase: str) -> bytes:
    return _gpg(["--symmetric", "--cipher-algo", "AES256", "--armor"], plain, passphrase)


def decrypt(cipher: bytes, passphrase: str) -> bytes:
    return _gpg(["--decrypt"], cipher, passphrase)

File: src/core/crypto.ts
/**
 * Symmetric encryption via gpg. The passphrase is handed to gpg over a
 * dedicated pipe fd (fd 3), never on argv (where `ps` would show it) and never
 * written to disk. Ciphertext/plaintext travel over stdin/stdout.
 */

import { spawn } from "node:child_process";
import { BroadcastError } from "../errors.js";

function gpg(args: string[], input: Buffer, passphrase: string): Promise<Buffer> {
  return new Promise((resolve, reject) => {
    const child = spawn(
      "gpg",
      [
        "--batch",
        "--yes",
        "--quiet",
        "--pinentry-mode",
        "loopback",
        "--passphrase-fd",
        "3",
        ...args,
      ],
      { stdio: ["pipe", "pipe", "pipe", "pipe"] },
    );
    const out: Buffer[] = [];
    const err: Buffer[] = [];
    child.stdout.on("data", (d: Buffer) => out.push(d));
    child.stderr.on("data", (d: Buffer) => err.push(d));
    child.on("error", (e) =>
      reject(
        new BroadcastError(e.message.includes("ENOENT") ? "gpg not found on PATH" : e.message),
      ),
    );
    child.on("close", (code) => {
      if (code === 0) resolve(Buffer.concat(out));
      else reject(new BroadcastError(`gpg failed: ${Buffer.concat(err).toString().trim()}`));
    });

    // fd 3 is the passphrase pipe; write it and close before feeding stdin.
    const passphraseStream = child.stdio[3] as NodeJS.WritableStream;
    passphraseStream.write(`${passphrase}\n`);
    passphraseStream.end();
    child.stdin.write(input);
    child.stdin.end();
  });
}

export function encrypt(plain: Buffer, passphrase: string): Promise<Buffer> {
  return gpg(["--symmetric", "--cipher-algo", "AES256", "--armor"], plain, passphrase);
}

export function decrypt(cipher: Buffer, passphrase: string): Promise<Buffer> {
  return gpg(["--decrypt"], cipher, passphrase);
}
